libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1055 @@
+      struct {                                                                                                         \
+         LIST_INDICES(value_class, __VA_ARGS__)                                                                        \
+      } index;                                                                                                         \
+                                                                                                                       \
+      table_class(eosio::name contract_name) {                                                                         \
+         init(contract_name, &index);                                                                                  \
+      }                                                                                                                \
+   };
+
+namespace eosio {
+   namespace internal_use_do_not_use {
+      extern "C" {
+         __attribute__((eosio_wasm_import))
+         void kv_erase(uint64_t db, uint64_t contract, const char* key, uint32_t key_size);
+
+         __attribute__((eosio_wasm_import))
+         void kv_set(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, const char* value, uint32_t value_size);
+
+         __attribute__((eosio_wasm_import))
+         bool kv_get(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, uint32_t& value_size);
+
+         __attribute__((eosio_wasm_import))
+         uint32_t kv_get_data(uint64_t db, uint32_t offset, char* data, uint32_t data_size);
+
+         __attribute__((eosio_wasm_import))
+         uint32_t kv_it_create(uint64_t db, uint64_t contract, const char* prefix, uint32_t size);
+
+         __attribute__((eosio_wasm_import))
+         void kv_it_destroy(uint32_t itr);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_status(uint32_t itr);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_compare(uint32_t itr_a, uint32_t itr_b);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_key_compare(uint32_t itr, const char* key, uint32_t size);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_move_to_end(uint32_t itr);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_next(uint32_t itr);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_prev(uint32_t itr);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_lower_bound(uint32_t itr, const char* key, uint32_t size);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_key(uint32_t itr, uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size);
+
+         __attribute__((eosio_wasm_import))
+         int32_t kv_it_value(uint32_t itr, uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size);
+      }
+   }
+
+/**
+ * The key_type struct is used to store the binary representation of a key.
+ */
+struct key_type {
```

Comment:
This type can simply inherit from string and retain type safety so conversion from a string can't be made implicitly.  We can also inherit from eosio::string and not std::string to save on WASM bloat generated from std::string.

Reply:
After the `Release` fix, I encountered a max call depth error. Switching to std::string leads to no errors, as well as no noticeable binary bloat, so I went with that.